File: 2024_semester_2/oop/lab7/app/model/BST.java
package model;

/**
 * Represents a Binary Search Tree (BST) data structure.
 * The BST class allows insertion, deletion, searching, and printing of elements in the tree.
 * It uses a generic type parameter T that must implement the Comparable interface.
 *
 * @param <T> the type of elements stored in the BST
 */
public class BST<T extends Comparable<T>> {
    Node<T> root;

    /**
     * Constructs an empty BST.
     */
    public BST() {
        this.root = null;
    };

    /**
     * Represents a node in the BST.
     * Each node has a value of type E, and may have a left child, right child, and parent node.
     *
     * @param <E> the type of value stored in the node
     */
    public class Node<E extends Comparable<E>> {
        E value;
        Node<E> left, right;

        public Node(E value) {
            this.value = value;
        }
    }

    /**
     * Inserts a value into the BST.
     *
     * @param value the value to insert
     */
    public void insert(T value) {
        System.out.println("Inserting value: " + value);
        if (root == null) {
            System.out.println("Inserting root");
            root = new Node<T>(value);
        } else {
            System.out.println("Inserting node");
            insertNode(root, value);
        }
    }

    /**
     * Deletes a value from the BST.
     *
     * @param value the value to delete
     */
    public void delete(T value) {
        root = deleteNode(root, value);
    }

    /**
     * Searches for a value in the BST.
     *
     * @param value the value to search for
     * @return true if the value is found, false otherwise
     */
    public boolean search(T value) {
        return searchRecursive(root, value);
    }

    /**
     * Prints the BST in a readable format.
     *
     * @return a string representation of the BST
     */
    public String print() {
        return printRecursive(root,0);
    }

    /**
     * Inserts a value into the BST recursively.
     * @param node the current node
     * @param value the value to insert
     */
    private void insertNode(Node<T> node, T value) {
        if (value.compareTo(node.value) < 0) {
            if (node.left == null) {
                node.left = new Node<T>(value);
            } else {
                insertNode(node.left, value);
            }
        } else {
            if (node.right == null) {
                node.right = new Node<T>(value);
            } else {
                insertNode(node.right, value);
            }
        }
    }

    /**
     * Deletes a value from the BST recursively.
     * @param node the current node
     * @param value the value to delete
     * @return the node after deletion
     */
    private Node<T> deleteNode(Node<T> node, T value) {
        if (node == null) {
            return null;
        }

        if (value.compareTo(node.value) < 0) {
            node.left = deleteNode(node.left, value);
        } else if (value.compareTo(node.value) > 0) {
            node.right = deleteNode(node.right, value);
        } else {
            if (node.left == null && node.right == null) {
                node = null;
            } else if (node.left == null) {
                node = node.right;
            } else if (node.right == null) {
                node = node.left;
            } else {
                Node<T> temp = node.right;
                while (temp.left != null) {
                    temp = temp.left;
                }
                node.value = temp.value;
                node.right = deleteNode(node.right, temp.value);
            }
        }

        return node;
    }

    /**
     * Searches for a value in the BST recursively.
     * @param node the current node
     * @param value the value to search for
     * @return true if the value is found, false otherwise
     */
    private boolean searchRecursive(Node<T> node, T value) {
        if (node == null) {
            return false;
        }
        if (value.compareTo(node.value) < 0) {
            return searchRecursive(node.left, value);
        } else if (value.compareTo(node.value) > 0) {
            return searchRecursive(node.right, value);
        } else {
            return true;
        }
    }

    /**
     * Prints the BST in a readable format recursively.
     * @param node the current node
     * @param indent the indentation level
     * @return a string representation of the BST
     */
    private String printRecursive(Node<T> node, int indent) {
        StringBuilder sb = new StringBuilder();
        if (node != null) {
            sb.append(printRecursive(node.right, indent + 4));
            for (int i = 0; i < indent; i++) {
                sb.append(" ");
            }
            sb.append("").append(node.value).append("\n");
            sb.append(printRecursive(node.left, indent + 4));
        }
        return sb.toString();
    }
}

File: 2024_semester_2/programming-course/lab7/cpp/src/bst.cpp
#include <iostream>
#include <string>
#include <sstream>

#include <bst.hpp>
#include <stringConverter.hpp>

template <typename T>
void BST<T>::insert(T value)
{
    std::cout << "Inserting value: " << value << std::endl;
    if (root == nullptr)
    {
        std::cout << "Inserting root" << std::endl;
        root = new Node(value);
    }
    else
    {
        std::cout << "Inserting node" << std::endl;
        insertNode(root, value);
    }
}

template <typename T>
void BST<T>::remove(T value)
{
    root = deleteNode(root, value);
}

template <typename T>
bool BST<T>::search(T value)
{
    return searchRecursive(root, value);
}

template <typename T>
std::string BST<T>::print()
{
    return printRecursive(root, 0);
}

template <typename T>
void BST<T>::insertNode(Node *node, T value)
{
    if (value < node->value)
    {
        if (node->left == nullptr)
        {
            node->left = new Node(value);
            node->left->parent = node;
        }
        else
        {
            insertNode(node->left, value);
        }
    }
    else
    {
        if (node->right == nullptr)
        {
            node->right = new Node(value);
            node->right->parent = node;
        }
        else
        {
            insertNode(node->right, value);
        }
    }
}

template <typename T>
typename BST<T>::Node *BST<T>::deleteNode(Node *node, T value)
{
    if (node == nullptr)
    {
        return nullptr;
    }

    if (value < node->value)
    {
        node->left = deleteNode(node->left, value);
    }
    else if (value > node->value)
    {
        node->right = deleteNode(node->right, value);
    }
    else
    {
        if (node->left == nullptr && node->right == nullptr)
        {
            delete node;
            node = nullptr;
        }
        else if (node->left == nullptr)
        {
            Node *temp = node;
            node = node->right;
            node->parent = temp->parent;
            delete temp;
        }
        else if (node->right == nullptr)
        {
            Node *temp = node;
            node = node->left;
            node->parent = temp->parent;
            delete temp;
        }
        else
        {
            Node *temp = node->right;
            while (temp->left != nullptr)
            {
                temp = temp->left;
            }
            node->value = temp->value;
            node->right = deleteNode(node->right, temp->value);
        }
    }

    return node;
}

template <typename T>
bool BST<T>::searchRecursive(Node *node, T value)
{
    if (node == nullptr)
    {
        return false;
    }
    if (value < node->value)
    {
        return searchRecursive(node->left, value);
    }
    else if (value > node->value)
    {
        return searchRecursive(node->right, value);
    }
    else
    {
        return true;
    }
}

template <typename T>
std::string BST<T>::printRecursive(Node *node, int indent)
{
    std::stringstream ss;
    
    if (node != nullptr)
    {
        ss << printRecursive(node->right, indent + 4);
        for (int i = 0; i < indent; i++)
        {
            ss << " ";
        }
        ss << StringConverter::valToString<T>(node->value) << "\n";
        ss << printRecursive(node->left, indent + 4);
    }

    return ss.str();
}

template void BST<int>::insert(int);
template void BST<int>::remove(int);
template bool BST<int>::search(int);
template std::string BST<int>::print();
template void BST<std::string>::insert(std::string);
template void BST<std::string>::remove(std::string);
template bool BST<std::string>::search(std::string);
template std::string BST<std::string>::print();
template void BST<double>::insert(double);
template void BST<double>::remove(double);
template bool BST<double>::search(double);
template std::string BST<double>::print();
